### DISCUSSION TITLE
portable-pty escape example on Windows

### DISCUSSION THREAD

**Discussion by jjangga0214** [wezterm/wezterm#2459] [Q&A]
Hi!

`portable-pty`'s example contains the comment below.

https://github.com/wez/wezterm/blob/9732f9ee2b1795d79ec7be1788045a3961fb55fd/pty/examples/whoami_async.rs#L60-L67

I could not understand what this means, literally.

> the windows conpty implementation synthesizes title change escape sequences in the output stream

What does this mean?
Would someone please explain it?
(e.g. by screenshot..?)

Thank you so much in advance.

**Comment by wez** [comment]
The posix PTY is a fairly lightweight stream device in the kernel, where data is passed between the application and the terminal emulator almost untouched; the kernel will look at the termios configuration to determine whether local echo is enabled, and whether certain control sequences such as VINTR are mapped to sending SIGINT.

The windows ConPTY is a bit more involved because it needs to support legacy win32 console APIs that provide access to a screen buffer.  The data flow is more complex: the output from the application is parsed by ConPTY and applied to the win32 screen buffer, and then that buffer is re-encoded and that output is then delivered to the terminal emulator.  The re-encoding can produce surprisingly verbose additional output, such as OSC codes that change the window title to match the running program and other mode changes.

3bf551706ea6353f2f1d0fa12ef3cf6d7cf7d3d7 (the commit that introduced that escaping) has this example output:

```
child status: ExitStatus { successful: true }
output: wez\r\n%
```

Example output on windows:

```
child status: ExitStatus { successful: true }
output: \u{1b}[2J\u{1b}[?25l\u{1b}[m\u{1b}[HDOMAIN\\wez\u{8}\u{1b}]0;C:\\windows\\system32\\whoami.EXE\u{0}\u{7}\u{1b}[?25h
```

It's undesirable for the sample program to pass those escapes right through because they do things like move the cursor location and change the title of terminal in which the example is run, which is rather surprising.
